mobile/lib/shared/services/sync.service.dart

```diff
@@ -0,0 +1,592 @@
+  Future<bool> syncUsersFromServer(List<User> users) async {
+    users.sortBy((u) => u.id);
+    final dbUsers = await _db.users.where().sortById().findAll();
+    final List<int> toDelete = [];
+    final List<User> toUpsert = [];
+    final changes = diffSortedListsSync(
+      users,
+      dbUsers,
+      compare: (User a, User b) => a.id.compareTo(b.id),
+      both: (User a, User b) {
+        if (a.updatedAt != b.updatedAt) {
+          toUpsert.add(a);
+          return true;
+        }
+        return false;
+      },
+      onlyFirst: (User a) => toUpsert.add(a),
+      onlySecond: (User b) => toDelete.add(b.isarId),
+    );
+    if (changes) {
+      await _db.writeTxn(() async {
+        await _db.users.deleteAll(toDelete);
+        await _db.users.putAll(toUpsert);
+      });
+    }
+    return changes;
+  }
+
+  /// Syncs remote assets owned by the logged-in user to the DB
+  /// Returns `true` if there were any changes
+  Future<bool> syncRemoteAssetsToDb(List<Asset> remote) async {
+    if (!_remoteAssetCompleter.isCompleted) {
+      return _remoteAssetCompleter.future;
+    }
+    _remoteAssetCompleter = Completer();
+    bool changes = false;
+    try {
+      changes = await _syncRemoteAssetsToDb(remote);
+    } finally {
+      _remoteAssetCompleter.complete(changes);
+    }
+    return changes;
+  }
+
+  /// Syncs remote albums to the database
+  /// returns `true` if there were any changes
+  Future<bool> syncRemoteAlbumsToDb(
+    List<AlbumResponseDto> remote, {
+    required bool isShared,
+    required FutureOr<AlbumResponseDto> Function(AlbumResponseDto) loadDetails,
+  }) async {
+    if (!_remoteAlbumCompleter.isCompleted) {
+      return _remoteAlbumCompleter.future;
+    }
+    _remoteAlbumCompleter = Completer();
+    bool changes = false;
+    try {
+      changes = await _syncRemoteAlbumsToDb(
+        remote,
+        isShared: isShared,
+        loadDetails: loadDetails,
+      );
+    } finally {
+      _remoteAlbumCompleter.complete(changes);
+    }
+    return changes;
+  }
+
+  /// Syncs all device albums and their assets to the database
+  /// Returns `true` if there were any changes
+  Future<bool> syncLocalAlbumAssetsToDb(List<AssetPathEntity> onDevice) async {
+    if (!_deviceCompleter.isCompleted) {
+      return _deviceCompleter.future;
+    }
+    _deviceCompleter = Completer();
+    bool changes = false;
+    try {
+      changes = await _syncLocalAlbumAssetsToDb(onDevice);
+    } finally {
+      _deviceCompleter.complete(changes);
+    }
+    return changes;
+  }
+
+  /// returns all Asset IDs that are not contained in the existing list
+  List<int> sharedAssetsToRemove(
+    List<Asset> deleteCandidates,
+    List<Asset> existing,
+  ) {
+    if (deleteCandidates.isEmpty) {
+      return [];
+    }
+    deleteCandidates.sort(Asset.compareById);
+    existing.sort(Asset.compareById);
+    return _diffAssets(existing, deleteCandidates, compare: Asset.compareById)
```

Comment:
Frist time seeing the usage of `Triple` pretty cool!!

Reply:
I hope tuple/record support becomes stable in Dart/Flutter soon. That we be a lot nicer than these custom types (pair, triple, ...)

Reply:
I think the Record type is available in Flutter 3.7?

Reply:
kind of. it is experimental and needs custom build switches to enable. I'd rather avoid that and wait for flutter 3.8 or whatever :-)